library/core/src/alloc/layout.rs

```diff
@@ -403,9 +405,17 @@ impl Layout {
     #[stable(feature = "alloc_layout_manipulation", since = "1.44.0")]
     #[inline]
     pub fn array<T>(n: usize) -> Result<Self, LayoutError> {
-        let (layout, offset) = Layout::new::<T>().repeat(n)?;
-        debug_assert_eq!(offset, mem::size_of::<T>());
-        Ok(layout.pad_to_align())
+        let array_size = mem::size_of::<T>().checked_mul(n).ok_or(LayoutError)?;
+
+        // SAFETY:
+        // - Size: `array_size` cannot be too big because `size_of::<T>()` must
+        //   be a multiple of `align_of::<T>()`. Therefore, `array_size`
+        //   rounded up to the nearest multiple of `align_of::<T>()` is just
+        //   `array_size`. And `array_size` cannot be too big because it was
+        //   just checked by the `checked_mul()`.
+        // - Alignment: `align_of::<T>()` will always give an acceptable
+        //   (non-zero, power of two) alignment.
```

Comment:
Is this true? I vaguely recall some conversation that this only applies to `repr(Rust)` types buy that some weird extern types could have non-power of two alignments. But I can't find a source on that, so I may be misremembering.

Reply:
This is fine according to https://doc.rust-lang.org/1.56.0/reference/type-layout.html#size-and-alignment, which doesn't say it's specific to repr(Rust).

> Alignment is measured in bytes, and must be at least 1, and always a power of 2. The alignment of a value can be checked with the `align_of_val` function.

Reply:
Nevermind, Layout itself requires it and so do C and C++, so I must have been thinking of something else.